From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Michael Shavit <mshavit@google.com>
Date: Thu, 20 Feb 2020 00:47:09 -0500
Subject: [PATCH] 1557: Fix an issue with ELFExternalSymbolResolver

Fix an issue where ExternalSymbolResolve doesn't resolve symbols that live in nested namespaces under <EXTERNAL>.
---
 .../program/util/ELFExternalSymbolResolver.java | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/Ghidra/Features/Base/src/main/java/ghidra/program/util/ELFExternalSymbolResolver.java b/Ghidra/Features/Base/src/main/java/ghidra/program/util/ELFExternalSymbolResolver.java
index c9563d1a3..8e8ee68da 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/program/util/ELFExternalSymbolResolver.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/program/util/ELFExternalSymbolResolver.java
@@ -148,7 +148,12 @@ public class ELFExternalSymbolResolver {
 				continue;
 			}
 			try {
-				s.setNamespace(extLibrary);
+				Symbol parent_symbol = s.getParentSymbol();
+				if (parent_symbol != null && parent_symbol.isExternal()) {
+					parent_symbol.setNamespace(extLibrary);
+				} else {
+					s.setNamespace(extLibrary);
+				}
 				idIterator.remove();
 				libResolvedCount++;
 				Msg.debug(ELFExternalSymbolResolver.class, "External symbol " + extLoc.getLabel() +
@@ -180,13 +185,9 @@ public class ELFExternalSymbolResolver {
 
 	private static Collection<Long> getUnresolvedExternalFunctionIds(Program program) {
 		List<Long> symbolIds = new ArrayList<>();
-		ExternalManager externalManager = program.getExternalManager();
-		Library library = externalManager.getExternalLibrary(Library.UNKNOWN);
-		if (library != null) {
-			for (Symbol s : program.getSymbolTable().getSymbols(library)) {
-				if (s.getSymbolType() == SymbolType.FUNCTION) {
-					symbolIds.add(s.getID());
-				}
+		for (Symbol s : program.getSymbolTable().getExternalSymbols()) {
+			if (s.getSymbolType() == SymbolType.FUNCTION) {
+				symbolIds.add(s.getID());
 			}
 		}
 		return symbolIds;
-- 
2.38.1

